File: MPI_Benchmark/source_code/pingpong/java/PingPong.java
import java.util.concurrent.Semaphore;

public class PingPong extends Thread{
    private Semaphore semaphore;

    private long timeSpawn;
    private long timeExec;

    public long timeStart;
    public long timeEnd;

    private int tamDados;
    private int qtdRept;
    private int pairsN;
    private String outLocation;

    public PingPong(int tamDados, int qtdMsg, int pairsN, String outLocation) {
        this.tamDados = tamDados;
        this.qtdRept = qtdMsg;
        this.pairsN = pairsN;
        this.outLocation = outLocation;
        this.semaphore = new Semaphore(-pairsN + 1);
    }

    public void run() {
        byte[] dado = generateData(tamDados);
        ProcPing[] procsPing = new ProcPing[pairsN];
        ProcPong[] procsPong = new ProcPong[pairsN];

        timeStart = timeMicroSeg();
        for (int i = 0; i < pairsN; i++) {
            procsPing[i] = new ProcPing("" + (2*i), dado, this, qtdRept);
            procsPong[i] = new ProcPong("" + (2*i+1), dado, qtdRept);
        }
        timeEnd = timeMicroSeg();
        timeSpawn = timeEnd - timeStart;

        timeStart = timeMicroSeg();
        for (int i = 0; i < pairsN; i++) {
            procsPing[i].setPeer(procsPong[i]);
            procsPong[i].setPeer(procsPing[i]);

            procsPing[i].start();
            procsPong[i].start();
        }
        dormirAteTerminar();
        timeEnd = timeMicroSeg();
        timeExec = timeEnd - timeStart;

        Salvar.writeResultPeer(outLocation, tamDados, qtdRept, timeExec, timeSpawn);
    }

    private void dormirAteTerminar() {
        try {
            semaphore.acquire();
        }
        catch (InterruptedException ie) {}
    }

    public void acordar() {
        semaphore.release();
    }

    private byte[] generateData(int tamDados) {
        byte[] dado = new byte[tamDados];
        return dado;
    }

    private long timeMicroSeg() {
        return System.nanoTime()/1000;
    }
}

File: MPI_Benchmark/source_code/pingpong/python/pingpong.py
#!/usr/bin/env python
# -*- coding: latin-1 -*-

import sys
import datetime
from threading import Thread
from threading import BoundedSemaphore

semaphore = BoundedSemaphore()
global_count = 0

class ProcPing(Thread):
	def __init__(self, name, data, qtdMsg):
		Thread.__init__(self)
		self.name = name
		self.data = data
		self.qtdMsg = qtdMsg
		self.mailBox = []

	def setPeer(self, Peer):
		self.Peer = Peer

	def send(self, dado):
		self.Peer.mailBox = dado

	def recv(self):
		while True:
			if not len(self.mailBox) < len(self.data):
				self.mailBox = []
				break

	def run(self):
		global global_count
		for i in range (0, self.qtdMsg + 1):
			self.send(self.data)
			if i < self.qtdMsg:
				self.recv()
		semaphore.acquire()
		global_count -= 1
		semaphore.release()

class ProcPong(Thread):
	def __init__(self, name, data, qtdMsg):
		Thread.__init__(self)
		self.name = name
		self.data = data
		self.qtdMsg = qtdMsg
		self.mailBox = []

	def setPeer(self, Peer):
		self.Peer = Peer

	def send(self, dado):
		self.Peer.mailBox = dado

	def recv(self):
		while True:
			if not len(self.mailBox) < len(self.data):
				self.mailBox = []
				break

	def run(self):
		global global_count
		for i in range (0, self.qtdMsg + 1):
			self.recv()
			if i < self.qtdMsg:
				self.send(self.data)
		semaphore.acquire()
		global_count -= 1
		semaphore.release()

class PingPing(Thread):

	def __init__(self, tamMsg, qtdMsg, PairsN):
		Thread.__init__(self)
		self.tamMsg = tamMsg
		self.qtdMsg = qtdMsg
		self.PairsN = PairsN

	def run(self):
		global global_count
		index = 0
		array = [1]
		pairs = []
		while index < self.tamMsg -1:
			array.append(1)
			index = index + 1

		timeStart = datetime.datetime.now()
		for pair in range(self.PairsN):
			p1 = ProcPing("1", array, self.qtdMsg)
			p2 = ProcPong("2", array, self.qtdMsg)
			pairs += [ (p1, p2), ]

		timeEnd = datetime.datetime.now()
		timeSpawn = timeEnd - timeStart

		timeStart = datetime.datetime.now()
		for p1, p2 in pairs:
			p2.setPeer(p1)
			p1.setPeer(p2)

			p1.start()
			p2.start()

		while global_count != 0:
			print global_count
			pass
		timeEnd = datetime.datetime.now()
		timeExec = timeEnd - timeStart

		line = "%d\t%d\t%s\t%s\n" % (self.tamMsg, self.qtdMsg, timeExec, timeSpawn)

		try:
			arq = open('saida.txt', 'r')
			textoSaida = arq.read()
			arq.close()
		except:
			arq = open('saida.txt', 'w')
			textoSaida = ""
			arq.close()

		arq = open('saida.txt', 'w')
		textoSaida = textoSaida + line
		arq.write(textoSaida)
		arq.close()

def main():
	param = sys.argv[1:]

	tamMsg = int(param[0])
	qtdMsg = int(param[1])
	PairsN = int(param[2])

	pingPing = PingPing(tamMsg, qtdMsg, PairsN)
	global global_count
	global_count = PairsN*2
	pingPing.start()

if __name__=="__main__":
	main()
